# Text Mining and ML example Example

We have a dataframe with tweets from both members of the Republican as well as the Democrat Party. We would like to analyse which party members cite more clikcbait articles. This could give an indication about the type of articles they or their voters like to consume. For this we first need to label the articles headers as clikbait or non clickbait using Logistic regression which we train on a dataset with article headers where we aleady know if they are clickbiat or not. Then we can compare the clikcbait artice use between democrats and republicans taking into account their total amount of tweets and look at potential factors why one party uses more clickbait articles than the other.



In [ ]:
#First I import all the necessary packages. In this case
# pandas to work with a dataframe
#pickle since we are dealing with a .p file and we need this package to upload such a file
#train_test_split to split our data into training and test data
#CountVectorizer to vectorize our text and to tokenize it which is necessary before performing logistic regression 
#from sklearn.linear_model import LogisticRegression to train our model to be able to receive a function that predicts the label (non-clickbait, clickbait) of our test set
#chi2_contingency to form a contigency table for a later chi square test


In [ ]:
import pandas as pd
import pickle
from sklearn.model_selection import train_test_split
from sklearn.feature_extraction.text import CountVectorizer
from sklearn.linear_model import LogisticRegression

!pip install scipy
from scipy.stats import chi2_contingency 



In [ ]:
#I ,then ,upload the framing. p file using pickle and define this as my dataframe. Here, I need to specify the path of the file as it is not my working directory.

df = pd.read_pickle(r"C:\Users\Luitwin\Desktop\Master Dat Science\Data Mining\week 2/framing.p")

In [ ]:
#For a first overvierw I print the dataframe.

df

tweet_id                date        user party  \
0      1325914751495499776 2020-11-09 21:34:45   SenShelby     R   
1      1294021087118987264 2020-08-13 21:20:43   SenShelby     R   
2      1323340848130609156 2020-11-02 19:06:59   DougJones     D   
3      1323004075831709698 2020-11-01 20:48:46   DougJones     D   
4      1322567531320717314 2020-10-31 15:54:06   DougJones     D   
...                    ...                 ...         ...   ...   
30803  1305999044196675584 2020-09-15 22:36:51  Liz_Cheney     R   
30804  1301257687049687041 2020-09-02 20:36:23  Liz_Cheney     R   
30805  1296922209890574337 2020-08-21 21:28:45  Liz_Cheney     R   
30806  1296082124970180613 2020-08-19 13:50:33  Liz_Cheney     R   
30807  1295913277621284864 2020-08-19 02:39:36  Liz_Cheney     R   

                  state              chamber  \
0               Alabama              Senator   
1               Alabama              Senator   
2               Alabama              Senator   
3               Alabama              Senator   
4               Alabama              Senator   
...                 ...                  ...   
30803  Wyoming At-Large  U.S. Representative   
30804  Wyoming At-Large  U.S. Representative   
30805  Wyoming At-Large  U.S. Representative   
30806  Wyoming At-Large  U.S. Representative   
30807  Wyoming At-Large  U.S. Representative   

                                                   tweet     news_mention  \
0      ICYMI – @BusinessInsider declared #Huntsville ...  businessinsider   
1      Great news! Today @mazda_toyota announced an a...                    
2      He’s already quitting on the folks of Alabama ...                    
3      I know you guys are getting bombarded with fun...                    
4      Well looky here folks, his own players don’t t...                    
...                                                  ...              ...   
30803  EXCLUSIVE: Liz Cheney Asks The DOJ To Investig...                    
30804  Hairpocrisy!\n#PelosiBlowout \n\nNancy Pelosi ...                    
30805  RT @Kredo0: JUST IN: Congressional Republicans...                    
30806  Liz Cheney slams John Kerry's DNC remarks: Sur...                    
30807  Thank you, Wyoming. Honored by your support - ...      AP_Politics   

                                           url_reference  \
0      https://www.businessinsider.com/personal-finan...   
1      https://pressroom.toyota.com/mazda-and-toyota-...   
2      https://apnews.com/article/c73f0dfe8008ebaf85e...   
3      https://secure.actblue.com/donate/djfs-close?r...   
4      https://slate.com/culture/2020/10/tommy-tuberv...   
...                                                  ...   
30803  https://dailycaller.com/2020/09/15/liz-cheney-...   
30804  https://www.washingtontimes.com/news/2020/sep/...   
30805  https://freebeacon.com/national-security/congr...   
30806  https://www.foxnews.com/politics/liz-cheney-jo...   
30807                    https://apnews.com/Election2020   

                        netloc  \
0      www.businessinsider.com   
1         pressroom.toyota.com   
2                   apnews.com   
3           secure.actblue.com   
4                    slate.com   
...                        ...   
30803          dailycaller.com   
30804  www.washingtontimes.com   
30805           freebeacon.com   
30806          www.foxnews.com   
30807               apnews.com   

                                                   title  \
0      The 10 best US cities to move to if you want t...   
1      Mazda and Toyota Further Commitment to U.S. Ma...   
2      Tuberville, Jones fight for Senate seat in Ala...   
3                                           I just gave!   
4      What Tommy Tuberville’s Former Auburn Players ...   
...                                                  ...   
30803  EXCLUSIVE: Liz Cheney Asks The DOJ To Investig...   
30804  Nancy Pelosi says she was ‘set up’ by San Fran...   
30805  Congressional Republic

In [ ]:
#As we can see the data is unlabeled in terms clickbait or not. We know that we alredy have dataframe that has 
#a online article titeling regarding being click bait or not(clickbait.csv)
#We will use this labelled dataframe to train a logistic regression which we can 
#then use to predict the clickbait labels for our framing.p file/tweets.


In [ ]:
#So let us upload the labeled file using the code from the manual and have a first look at it by printing it. 
DATASET_URL = 'https://gist.githubusercontent.com/amitness/0a2ddbcb61c34eab04bad5a17fd8c86b/raw/66ad13dfac4bd1201e09726677dd8ba8048bb8af/clickbait.csv'
data = pd.read_csv(DATASET_URL)
data

title  label
0      15 Highly Important Questions About Adulthood,...      1
1      250 Nuns Just Cycled All The Way From Kathmand...      1
2      Australian comedians "could have been shot" du...      0
3      Lycos launches screensaver to increase spammer...      0
4      Fußball-Bundesliga 2008–09: Goalkeeper Butt si...      0
...                                                  ...    ...
31981   The Online Ad That Knows Where Your Friends Shop      0
31982         What Do You Love Based On Your Zodiac Sign      1
31983  This Baby's Reaction To Hearing About How She ...      1
31984  Saturn's moon Titan hosts liquid lakes and rivers      0
31985  Here's What Happened When People Tried Chicken...      1

[31986 rows x 2 columns]

In [ ]:
#Now can split this labelled dataframe in train test data. We choose 20 % test since 
#this has already produced an accurate model as seen in the manual. 

In [ ]:
X = list(data.title.values)
y = list(data.label.values)# the labels we want to predict --> Y
labels = ['not clickbait', 'clickbait']

X_train_str, X_test_str, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=1)

In [ ]:
#We use the Bag Of Words (BOW) to vectorize the text. We do not remove the stop words 
#since we have seen that some of them will later be classified as significant in classifying 
#a clickbait title. (E.g You and he)

In [ ]:
cv = CountVectorizer() # this initializes the CountVectorizer 

cv.fit(X_train_str) # create the vocabulary

X_train = cv.transform(X_train_str)
X_test = cv.transform(X_test_str)

In [ ]:
#We train the model with the vectorized text.

In [ ]:
from sklearn.linear_model import LogisticRegression

lr = LogisticRegression(solver='lbfgs')
lr.fit(X_train, y_train)

LogisticRegression()

In [ ]:
#we let it predict the cklickbait labels for each title from the test sest using predict()

In [ ]:
from sklearn.metrics import classification_report

y_pred = lr.predict(X_test)

In [ ]:
#We print out the accuracy and precision and the proportion of correct labels for each label category. 
print(classification_report(y_test, y_pred, 
                          target_names=labels))

               precision    recall  f1-score   support

not clickbait       0.96      0.98      0.97      3178
    clickbait       0.98      0.96      0.97      3220

     accuracy                           0.97      6398
    macro avg       0.97      0.97      0.97      6398
 weighted avg       0.97      0.97      0.97      6398



In [ ]:
#Another way to look at the performance of a classification model, is the so-called confusion matrix.
#Here, we can see how often a label is mistaken for another label

In [ ]:
from sklearn.metrics import confusion_matrix

print(confusion_matrix(y_test, y_pred))

[[3112   66]
 [ 115 3105]]


In [ ]:
#I look at the words that are very typical for clickbait titles. 
#Good that we didnt filter out the stop words.Some of them are very typical for clickbait titles (e.g you,we, her)

In [ ]:
vocabulary = cv.get_feature_names()
regression_coefficients = lr.coef_[0] # get the LR weights
vocab_coef_combined = list(zip(regression_coefficients, vocabulary)) # this combines two separate lists [1, 2], ['word1', 'word2'] into one list [[1, 'word1'], [2, 'word2']]

feature_importance = pd.DataFrame(vocab_coef_combined,
                      columns=['coef', 'word'])
feature_importance.sort_values('coef', ascending=False).head(60)

coef         word
20545  4.018000          you
18566  3.058204        these
18594  2.929925         this
154    2.922905         2015
2962   2.900514     buzzfeed
168    2.783366           21
8694   2.774251         here
12329  2.670787           my
18575  2.663407       things
20553  2.609577         your
4163   2.567400  confessions
75     2.565462           17
6519   2.538644     everyone
90     2.363792           19
20037  2.330103         ways
20043  2.325280           we
63     2.291693           15
20143  2.287868         what
20195  2.251463          why
8994   2.244905          how
176    2.243084           23
6516   2.228951        every
83     2.195795           18
13605  2.185682       people
52     2.153448           13
155    2.143770         2016
8380   2.124998    halloween
13624  2.110969      perfect
1164   2.108635       anyone
8295   2.062117         guys
172    2.047235           22
641    2.032608     adorable
8688   1.985482          her
8036   1.968439          got
198    1.942841           27
586    1.929518     actually
19112  1.914154          try
19215  1.872145      twitter
2122   1.866890         best
19889  1.858416           vs
20160  1.855602        which
17226  1.844711      someone
20003  1.834385        watch
69     1.831475           16
59     1.825435           14
20351  1.817125        women
18564  1.806590        there
10050  1.805114         just
8760   1.804809    hilarious
8349   1.791481         hair
20184  1.776408          who
8249   1.766696        guess
13770  1.751287       photos
9513   1.750681    instagram
204    1.736372           29
17823  1.736236    struggles
8329   1.733493          had
14970  1.727446         read
16752  1.718285       should
15002  1.706167      reasons

In [ ]:
#Since we have a model with a high accuracy
#we can now predict the labels of the titles in the framing.p data frame.

In [ ]:

input = cv.transform(df['title']) #The title of our unlabelled df will be used as input 
y_pred = lr.predict(input) #to then predict the labels using the model we trained before


In [ ]:
y_pred #lets print them to see if it went well

array([1, 0, 0, ..., 0, 0, 0], dtype=int64)

In [ ]:
#We take the output and attach to the df unlabelled data frame. 

In [ ]:
df['labelclickbait'] = y_pred

In [ ]:
#Let us look at the df to see if the attachment worked. The new labels have been sucessfully added as column 'labelclickbait'.

In [ ]:
df.head(60) 

tweet_id                date       user party  \
0   1325914751495499776 2020-11-09 21:34:45  SenShelby     R   
1   1294021087118987264 2020-08-13 21:20:43  SenShelby     R   
2   1323340848130609156 2020-11-02 19:06:59  DougJones     D   
3   1323004075831709698 2020-11-01 20:48:46  DougJones     D   
4   1322567531320717314 2020-10-31 15:54:06  DougJones     D   
6   1320853775893094401 2020-10-26 22:24:15  DougJones     D   
8   1319698596074262530 2020-10-23 17:53:58  DougJones     D   
9   1319690481501114368 2020-10-23 17:21:44  DougJones     D   
10  1319090524612141061 2020-10-22 01:37:43  DougJones     D   
11  1318544465146109954 2020-10-20 13:27:52  DougJones     D   
12  1317670841480609792 2020-10-18 03:36:24  DougJones     D   
14  1317220978951360517 2020-10-16 21:48:48  DougJones     D   
15  1316560425966866434 2020-10-15 02:04:00  DougJones     D   
17  1314405223847297025 2020-10-09 03:20:00  DougJones     D   
18  1311664519807209472 2020-10-01 13:49:25  DougJones     D   
19  1311091680402079745 2020-09-29 23:53:10  DougJones     D   
20  1310409838057451520 2020-09-28 02:43:46  DougJones     D   
21  1309636079172177925 2020-09-25 23:29:07  DougJones     D   
22  1308111571630129158 2020-09-21 18:31:16  DougJones     D   
23  1304221592185077763 2020-09-11 00:53:53  DougJones     D   
24  1303170536202285056 2020-09-08 03:17:22  DougJones     D   
25  1302628562055421953 2020-09-06 15:23:45  DougJones     D   
26  1300188176036790272 2020-08-30 21:46:32  DougJones     D   
27  1298054522821017600 2020-08-25 00:28:09  DougJones     D   
28  1295844901112942595 2020-08-18 22:07:54  DougJones     D   
30  1295517455436193792 2020-08-18 00:26:45  DougJones     D   
31  1295065848273670144 2020-08-16 18:32:14  DougJones     D   
32  1294779860607471616 2020-08-15 23:35:49  DougJones     D   
33  1293907842932432899 2020-08-13 13:50:44  DougJones     D   
34  1327314642163601409 2020-11-13 18:17:25   RepByrne     R   
35  1327314453369663488 2020-11-13 18:16:40   RepByrne     R   
36  1326331012327026689 2020-11-11 01:08:50   RepByrne     R   
37  1323632990229041152 2020-11-03 14:27:51   RepByrne     R   
38  1323345133086085121 2020-11-02 19:24:00   RepByrne     R   
39  1321520316569604101 2020-10-28 18:32:50   RepByrne     R   
40  1321206491299368963 2020-10-27 21:45:49   RepByrne     R   
42  1321088212052430849 2020-10-27 13:55:49   RepByrne     R   
43  1320859211702718464 2020-10-26 22:45:51   RepByrne     R   
44  1320437163562045444 2020-10-25 18:48:46   RepByrne     R   
46  1319702771092934657 2020-10-23 18:10:34   RepByrne     R   
47  1318637483748478977 2020-10-20 19:37:29   RepByrne     R   
48  1318598204972322817 2020-10-20 17:01:25   RepByrne     R   
49  1316123597027581954 2020-10-13 21:08:12   RepByrne     R   
50  1316084474338316297 2020-10-13 18:32:45   RepByrne     R   
53  1311807678231072768 2020-10-01 23:18:17   RepByrne     R   
54  1311728810199601152 2020-10-01 18:04:53   RepByrne     R   
55  1311688942744461318 2020-10-01 15:26:28   RepByrne     R   
56  1311397368479330306 2020-09-30 20:07:51   RepByrne     R   
57  1310909946830221313 2020-09-29 11:51:01   RepByrne     R   
59  1309552555312193536 2020-09-25 17:57:14   RepByrne     R   
60  1309230781974315012 2020-09-24 20:38:37   RepByrne     R   
61  1309115831830355969 2020-09-24 13:01:51   RepByrne     R   
62  1308750887758761984 2020-09-23 12:51:41   RepByrne     R   
63  1308123990049861634 2020-09-21 19:20:37   RepByrne     R   
64  1308123838488752129 2020-09-21 19:20:01   RepByrne     R   
66  1307384687808126978 2020-09-19 18:22:54   RepByrne     R   
71  1306354370288680962 2020-09-16 22:08:47   RepByrne     R   
73  1306294162287980547 2020-09-16 18:09:32   RepByrne     R   
74  1306247288898154496 2020-09-16 15:03:17   RepByrne     R   
75  1305919369877377029 2020-09-15 17:20:15   RepByrne     R   

                   state         chamber  \
0                Alabama         Senator   
1                Alaba

In [ ]:
df.tail(60)

tweet_id                date             user party  \
30735  1306367488368824321 2020-09-16 23:00:55     TomTiffanyWI     R   
30736  1305919332514496513 2020-09-15 17:20:06     TomTiffanyWI     R   
30737  1305189802208301057 2020-09-13 17:01:12     TomTiffanyWI     R   
30738  1304789467589378048 2020-09-12 14:30:25     TomTiffanyWI     R   
30740  1303717427910504448 2020-09-09 15:30:31     TomTiffanyWI     R   
30741  1302360873969881091 2020-09-05 21:40:03     TomTiffanyWI     R   
30742  1302296433077506050 2020-09-05 17:23:59     TomTiffanyWI     R   
30743  1301530427304312833 2020-09-03 14:40:09     TomTiffanyWI     R   
30744  1300063396012974080 2020-08-30 13:30:42     TomTiffanyWI     R   
30746  1297979322259644417 2020-08-24 19:29:20     TomTiffanyWI     R   
30747  1297571980049358848 2020-08-23 16:30:42     TomTiffanyWI     R   
30748  1297179438673793027 2020-08-22 14:30:53     TomTiffanyWI     R   
30749  1296460768553193473 2020-08-20 14:55:08     TomTiffanyWI     R   
30750  1296110866602037248 2020-08-19 15:44:45     TomTiffanyWI     R   
30751  1295774289258520576 2020-08-18 17:27:19     TomTiffanyWI     R   
30752  1294280361267077123 2020-08-14 14:30:59     TomTiffanyWI     R   
30754  1323706522166243329 2020-11-03 19:20:02        MikeforWI     R   
30755  1323251560831455232 2020-11-02 13:12:11        MikeforWI     R   
30757  1320070872087072768 2020-10-24 18:33:16        MikeforWI     R   
30759  1319046070383640577 2020-10-21 22:41:04        MikeforWI     R   
30760  1318586502608191488 2020-10-20 16:14:55        MikeforWI     R   
30761  1317862755228127233 2020-10-18 16:19:00        MikeforWI     R   
30762  1305605538366083073 2020-09-14 20:33:11        MikeforWI     R   
30763  1296602218016264193 2020-08-21 00:17:13        MikeforWI     R   
30764  1319330836173434880 2020-10-22 17:32:37      SenatorEnzi     R   
30765  1308837435128897543 2020-09-23 18:35:36      SenatorEnzi     R   
30766  1308837412550963200 2020-09-23 18:35:30      SenatorEnzi     R   
30767  1306652905416454144 2020-09-17 17:55:03      SenatorEnzi     R   
30768  1301608289013313537 2020-09-03 19:49:33      SenatorEnzi     R   
30769  1300468295599042562 2020-08-31 16:19:37      SenatorEnzi     R   
30770  1299463494115487745 2020-08-28 21:46:54      SenatorEnzi     R   
30771  1298723241284927488 2020-08-26 20:45:24      SenatorEnzi     R   
30772  1297991562387812353 2020-08-24 20:17:58      SenatorEnzi     R   
30773  1296081370318549003 2020-08-19 13:47:33      SenatorEnzi     R   
30774  1319822554811539458 2020-10-24 02:06:32  SenJohnBarrasso     R   
30776  1309157440357904392 2020-09-24 15:47:11  SenJohnBarrasso     R   
30777  1304039245074624512 2020-09-10 12:49:18  SenJohnBarrasso     R   
30778  1300466126896402433 2020-08-31 16:11:00  SenJohnBarrasso     R   
30779  1298298075962839040 2020-08-25 16:35:57  SenJohnBarrasso     R   
30780  1298285671820472320 2020-08-25 15:46:39  SenJohnBarrasso     R   
30782  1298059727415447553 2020-08-25 00:48:50  SenJohnBarrasso     R   
30783  1297970212223016961 2020-08-24 18:53:08  SenJohnBarrasso     R   
30784  1296435951858864128 2020-08-20 13:16:32  SenJohnBarrasso     R   
30785  1296113005248339969 2020-08-19 15:53:15  SenJohnBarrasso     R   
30786  1294645816196562944 2020-08-15 14:43:10  SenJohnBarrasso     R   
30787  1324364432315658240 2020-11-05 14:54:20       Liz_Cheney     R   
30788  1324191607302868992 2020-11-05 03:27:36       Liz_Cheney     R   
30789  1323829162721894400 2020-11-04 03:27:22       Liz_Cheney     R   
30790  1320906530103984136 2020-10-27 01:53:52       Liz_Cheney     R   
30792  1319083093664301056 2020-10-22 01:08:11       Liz_Cheney     R   
30798  1311231901299101696 2020-09-30 09:10:21       Liz_Cheney     R   
30799  1310966996499156992 2020-09-29 15:37:43       Liz_Cheney     R   
30800  1310605869663612935 2020-09-28 15:42:43       Liz_Cheney     R   
30801  1307314736128811017 2020-09-19 13:44:56       Liz_Cheney     R   
308

In [ ]:
#How many times do democrats refer to an article with a clickbait title and how many times do republicans do? 

#For that we look at the dataframe that shows clickbait titles used by the Democratic Party. We count the rows to the get the number.

In [ ]:
countD1 = df[(df['labelclickbait'] == 1) & (df['party'] =='D' )].shape[0]
countD1

2830

In [ ]:
#we get 2830 rows where democrats refer to an article with a title labelled by us a clickbait.
#Now we count the same for Republicans.

In [ ]:
countR1=df[(df['labelclickbait'] == 1) & (df['party'] =='R' )].shape[0]
countR1

1058

In [ ]:
#we get 1058 rows where republicans refer to an article with a title labelled by us a clickbait.

In [ ]:
CounttotalR = df[df['party'] =='R'].shape[0]
CounttotalD = df[df['party'] =='D'].shape[0]


In [ ]:
#lets look at the percentage as compared to total amounts of tweets per party.

In [ ]:
PercD = (countD1/CounttotalD)*100
PercD

18.182986378822925

In [ ]:
PercR = (countR1/CounttotalR)*100
PercR

13.61472140007721

In [ ]:
#We can also see that percentage wise (clickbait title of all post by this party) the Democrats refer to clickbait articles  more tham Republicans.Namely 4% percent more. 

In [ ]:
#To see if this difference is statistically significant we can do a CHi Square test to see if this difference is statistically significant.
#We first drop the I and L parrties from the dataframe, then make a contgency table based upon which we perform our chi square test.

In [ ]:
partyclickbaitcrosstab = pd.crosstab(df[(df['party'] =='D') | (df['party'] =='R' )]['labelclickbait'],
                                     df[(df['party'] =='D') | (df['party'] =='R' )]['party'], margins = False)
partyclickbaitcrosstab 

party               D     R
labelclickbait             
0               12734  6713
1                2830  1058

In [ ]:
#Using this table we can now perform this Chi Square Test. 

In [ ]:
stat, p, dof, expected = chi2_contingency(partyclickbaitcrosstab) 
 

In [ ]:
alpha = 0.05
print("p value is " + str(p)) 
if p <= alpha: 
    print('Dependent (reject H0)') 
else: 
    print('Independent (H0 holds true)') 

p value is 1.2810427655490298e-18
Dependent (reject H0)


In [ ]:
#the null hypothesis is that there is no significant difference between proportions in the labelled posts of Democarats and Republicans. This can be rejected and supports that this difference is statiscally significant. 

In [ ]:

dfclickbait = df[df['labelclickbait'] == 1] #allthe clickbait articles for a brief look 
dfclickbait

tweet_id                date             user party  \
0      1325914751495499776 2020-11-09 21:34:45        SenShelby     R   
3      1323004075831709698 2020-11-01 20:48:46        DougJones     D   
8      1319698596074262530 2020-10-23 17:53:58        DougJones     D   
9      1319690481501114368 2020-10-23 17:21:44        DougJones     D   
19     1311091680402079745 2020-09-29 23:53:10        DougJones     D   
...                    ...                 ...              ...   ...   
30771  1298723241284927488 2020-08-26 20:45:24      SenatorEnzi     R   
30776  1309157440357904392 2020-09-24 15:47:11  SenJohnBarrasso     R   
30786  1294645816196562944 2020-08-15 14:43:10  SenJohnBarrasso     R   
30787  1324364432315658240 2020-11-05 14:54:20       Liz_Cheney     R   
30801  1307314736128811017 2020-09-19 13:44:56       Liz_Cheney     R   

                  state              chamber  \
0               Alabama              Senator   
3               Alabama              Senator   
8               Alabama              Senator   
9               Alabama              Senator   
19              Alabama              Senator   
...                 ...                  ...   
30771           Wyoming         U.S. Senator   
30776           Wyoming         U.S. Senator   
30786           Wyoming         U.S. Senator   
30787  Wyoming At-Large  U.S. Representative   
30801  Wyoming At-Large  U.S. Representative   

                                                   tweet     news_mention  \
0      ICYMI – @BusinessInsider declared #Huntsville ...  businessinsider   
3      I know you guys are getting bombarded with fun...                    
8      We are excited to have @DebraMessing, @SeanHay...                    
9      Try this link:    https://t.co/TuqKYnXhsX http...                    
19     Proud to have the endorsement of my friend, th...                    
...                                                  ...              ...   
30771  RT @TRUCKINGdotORG: Next week, four female @Am...                    
30776  RT @Interior: #OTD in 1906, President Theodore...                    
30786  Wrote in @FoxNews about the coming #coronaviru...          FoxNews   
30787  RT @cnsnews: “The surge of victorious pro-life...          cnsnews   
30801  Must read from my friend and colleague @RepBri...                    

                                           url_reference  \
0      https://www.businessinsider.com/personal-finan...   
3      https://secure.actblue.com/donate/djfs-close?r...   
8            http://secure.actblue.com/donate/10.23.20wg   
9           https://secure.actblue.com/donate/10.23.20wg   
19                                       http://vote.org   
...                                                  ...   
30771  https://www.trucking.org/news-insights/louisa-...   
30776                  http://www.sharetheexperience.org   
30786  https://www.foxnews.com/opinion/coronavirus-va...   
30787  https://www.cnsnews.com/article/washington/pat...   
30801  https://www.foxnews.com/opinion/brian-mast-10t...   

                           netloc  \
0         www.businessinsider.com   
3              secure.actblue.com   
8              secure.actblue.com   
9              secure.actblue.com   
19                       vote.org   
...                           ...   
30771            www.trucking.org   
30776  www.sharetheexperience.org   
30786             www.foxnews.com   
30787             www.cnsnews.com   
30801             www.foxnews.com   

                                                   title  \
0      The 10 best US cities to move to if you want t...   
3                                           I just gave!   
8                                           I just gave!   
9                                           I just gave!   
19                Everything You Need to Vote - Vote.org   
...                                                  ...   
30771  Louisa Swain and the Trailblazers Who Show Us 

In [ ]:
dfclickbait.shape[0] #the amount of clickbait articles

3909

In [ ]:
dfD =df[df['party'] == 'D'] 
dfD.shape[0] #the amount of titles for the Democrats

15564

In [ ]:
dfR=df[df['party'] == 'R'] 
dfR.shape[0] #the amount of titles for the Republicans

7771

In [ ]:
dfclickbaitD =dfclickbait[dfclickbait['party'] == 'D'] #all the democrat clickbait titles
dfclickbaitD
dfclickbaitD.shape[0] 

2830

In [ ]:
dfclickbaitR =dfclickbait[dfclickbait['party'] == 'R'] #all the republicans clickbait titles
dfclickbaitR 
dfclickbaitR.shape[0]

1058

In [ ]:
#we nwo go thorugh some popular topics to see how they compare for each party

In [ ]:
dfclickbaitD[dfclickbaitD['title'].str.contains('Covid', na = False)].shape[0] 

118

In [ ]:
dfclickbaitD[dfclickbaitD['title'].str.contains('Corona', na = False)].shape[0]

44

In [ ]:
dfclickbaitR[dfclickbaitR['title'].str.contains('Covid', na = False)].shape[0]

9

In [ ]:
dfclickbaitR[dfclickbaitR['title'].str.contains('Corona', na = False)].shape[0]

32

In [ ]:
dfclickbaitD[dfclickbaitD['title'].str.contains('just gave', na = False)].shape[0]

23

In [ ]:
dfclikbaitR[dfclickbaitR['title'].str.contains('just gave', na = False)].shape[0]

1

In [ ]:
dfclickbaitD[dfclickbaitD['title'].str.contains('vote', na = False)].shape[0]

77

In [ ]:
dfclickbaitR[dfclickbaitR['title'].str.contains('vote', na = False)].shape[0]

22

In [ ]:
#lets look at the news sources used by the different parties and their clickbait proportions

In [ ]:
df.netloc.value_counts()

www.nytimes.com           1508
                          1092
www.politico.com           623
www.cnn.com                602
thehill.com                550
                          ... 
immigrantreliefwa.org        1
winningforwomen.com          1
help304.com                  1
VoteInNH.org                 1
allianceportregion.com       1
Name: netloc, Length: 3602, dtype: int64

In [ ]:
dfclickbait.netloc.value_counts()

www.nytimes.com                   312
www.nbcnews.com                   132
www.cnn.com                       117
www.theatlantic.com               114
www.foxnews.com                   106
                                 ... 
wrex.com                            1
www.yaleclimateconnections.org      1
northdallasgazette.com              1
spectrumnews1.com                   1
www.tnjustice.org                   1
Name: netloc, Length: 911, dtype: int64

In [ ]:
dfD.netloc.value_counts()

www.nytimes.com               1420
                               575
www.cnn.com                    563
www.politico.com               555
www.nbcnews.com                456
                              ... 
allinwa.org                      1
www.signupgenius.com             1
festival.texastribune.org        1
www.washingtoninformer.com       1
www.pinknews.co.uk               1
Name: netloc, Length: 2378, dtype: int64

In [ ]:
dfR.netloc.value_counts()

www.foxnews.com                  535
                                 511
thehill.com                      186
www.cnbc.com                     146
www.foxbusiness.com              126
                                ... 
cutoutoverdoses.com                1
www.nam.org                        1
hellowoodlands.com                 1
www.mcknightsseniorliving.com      1
www.falloncountyextra.com          1
Name: netloc, Length: 1727, dtype: int64

In [ ]:
dfclickbaitD.netloc.value_counts()

www.nytimes.com                     302
www.nbcnews.com                     128
www.theatlantic.com                 110
www.cnn.com                         107
nyti.ms                              77
                                   ... 
rssfeeds.tennessean.com               1
DemConvention.com                     1
www.hrc.org                           1
Ballotform.Philadelphiavotes.com      1
www.hachettebookgroup.com             1
Name: netloc, Length: 643, dtype: int64

In [ ]:
dfclickbaitR.netloc.value_counts()

www.foxnews.com           103
thehill.com                48
www.nationalreview.com     32
fxn.ws                     30
podcasts.apple.com         27
                         ... 
thetandd.com                1
bigten.org                  1
www.dcourier.com            1
cs.pn                       1
t.uga.edu                   1
Name: netloc, Length: 367, dtype: int64

# SUMMARY

We have seen that using a logistic regression as classifier that Democrats refer to Clickbait inn 18 % percent of all their tweets whereas Republicans do so only for 14%. 
Moreover, the difference Is significant. 
In total numbers Democrats have 1722 more clickbait title. Intuitively, one might that think that republicans refer more to clickbait articles. 

Whymore intuitive? Because the stemvoters of the republicanparty are majorly working class and thus often less educated. Classic clickbait titles such as: “Find out how you can make 4000 Dollars from home” might be more attractive to less educated people.

A look at the first 60 rows and titles of our newly labelled data frames gives us a hint why democrats could proprtionally more cb articles. There are many clickbait articles by the democrats with the same title. “I just gave”. This refers to a donation campaign by the democrats after Justice Ruth Bader Ginsburg’s death was announced where they reached a record-number of donations, as seen here also with the help of twitter.

https://www.nytimes.com/2020/09/18/us/elections/democrats-shatter-actblues-donation-records-after-ginsburgs-death.html

In fact if we look at the amount of clickbait titles with this type among democrats, we can see that they constitute for 23 clickbait titles among democrats
There is also a striking difference between Clickbait articles related to the Coronavirus between  Democrats and Republicans. This is line with reports suggesting that Democrats care more about the Coronavirus than Republicans
Democats also gave much more Clickbait insructions on “how to vote”.

Lastly,we saw that New York Times which is mainly cited by the Democrats accounts for 8% and the most of all clickbait articles (312 of a total of 3909 clickbait titles)which could be a main driver for significant difference of clickbait titles proportians between the two parties. A similar pattern can be obeserved for CNN. Politico.com a popular news source by the Republicans does not seem be very clickbaity.



